### PR TITLE
Feature 6 functionality implemented

### PR DESCRIPTION
**Pull request by nojuord**

### CODE REVIEW

File: src/functions/list-projects/handler.ts [hunk 1]
```diff
@@ -1,44 +1,27 @@
...
         return false;
       }
     } else if (project.start_date == null || currentDate <= new Date(project.start_date)) {
       return false;
     }
 
-    if (endDate) {
-        if (project.end_date == null || new Date(endDate) >= new Date(project.end_date)) {
+    if (parameters.endDate) {
+        if (project.end_date == null || new Date(parameters.endDate) >= new Date(project.end_date)) {
```
Comment (Jdallos) [inline src/functions/list-projects/handler.ts:24]: fix indentation

File: src/functions/list-tasks/handler.ts [hunk 2]
```diff
@@ -1,40 +1,78 @@
...
+  projectId: number,
+  remaining: number,
+  estimate: number,
+  startDate: string,
+  endDate: string,
+  highPriority: boolean,
+  assignedPersons: number[],
+}
+
+async function listTasksFunction(api: ForecastApiService, parameters: Parameters): Promise<Response[]> {
```
Comment (Jdallos) [inline src/functions/list-tasks/handler.ts:23]: function comments

File: src/functions/list-tasks/handler.ts [hunk 3]
```diff
@@ -1,40 +1,78 @@
...
+  if (parameters.projectId != null) {
```
Comment (Jdallos) [inline src/functions/list-tasks/handler.ts:24]:
```suggestion
  if (parameters.projectId !== null) {
```

File: src/functions/list-tasks/handler.ts [hunk 4]
```diff
@@ -1,40 +1,78 @@
...
+    const tasks = await api.getTasks(parameters.projectId);
+
+    var filteredTasks = tasks.filter(task => task.project_id == parameters.projectId);
```
Comment (Jdallos) [inline src/functions/list-tasks/handler.ts:27]: All var should be const, (unless they have to be let)

File: src/functions/list-tasks/handler.ts [hunk 5]
```diff
@@ -1,40 +1,78 @@
...
+  } else {
+    var filteredTasks = await api.getAllTasks();
+  }
+
+  const responseTasks = filteredTasks.map(task => {
```
Comment (Jdallos) [inline src/functions/list-tasks/handler.ts:32]: can be implicit return

File: src/functions/list-tasks/handler.ts (hunk 4 above)
Comment (Jdallos) [inline src/functions/list-tasks/handler.ts:27]:
```suggestion
    var filteredTasks = tasks.filter(task => task.project_id === parameters.projectId);
```

File: src/functions/list-time-entries/handler.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
+import { ValidatedEventAPIGatewayProxyEvent } from '@libs/api-gateway';
+// import { parseBearerAuth } from '@libs/auth-utils';
+import { middyfy } from '@libs/lambda';
+import { CreateForecastApiService, ForecastApiService } from 'src/apis/forecast-api-service';
+
+interface Parameters {
```
Comment (Jdallos) [inline src/functions/list-time-entries/handler.ts:6]: add comments

File: src/functions/list-time-entries/handler.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  projectId: number
+}
+
+interface Response {
```
Comment (Jdallos) [inline src/functions/list-time-entries/handler.ts:10]: add comments

File: src/functions/list-time-entries/handler.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  id: number,
+  person: number,
+  project: number,
+  task: number,
+  timeRegistered: number;
+}
+
+async function listTimeEntriesFunction(api: ForecastApiService, parameters: Parameters): Promise<Response[]> {
```
Comment (Jdallos) [inline src/functions/list-time-entries/handler.ts:18]: add comments

File: src/functions/list-time-entries/handler.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  const timeEntries = await api.getTimeEntries(parameters.projectId);
+
+  const responseTimeEntries = timeEntries.map(timeEntry => {
```
Comment (Jdallos) [inline src/functions/list-time-entries/handler.ts:21]: implicit return

**Review by Jdallos — COMMENTED** [review]
Some suggestions

File: src/functions/list-project-sprints/handler.ts [hunk 10]
```diff
@@ -58,7 +69,7 @@ const listProjectSprints: ValidatedEventAPIGatewayProxyEvent<any> = async event
   const api = CreateForecastApiService();
 
   const projectSprints = await listProjectSprintsFunction(api, {
-    projectId: +event.queryStringParameters.projectId
+    projectId: parseInt(event.queryStringParameters.projectId),
```
Comment (Jdallos) [inline src/functions/list-project-sprints/handler.ts:72]:
```suggestion
    projectId: parseInt(event.queryStringParameters.projectId)
```

File: src/functions/list-project-sprints/handler.ts [hunk 11]
```diff
@@ -1,34 +1,45 @@
...
+ * @returns Array of sprints
+ */
+async function listProjectSprintsFunction(api: ForecastApiService, parameters: ListProjectSprintsParameters): Promise<Response[]> {
   const sprints = await api.getProjectSprints(parameters.projectId);
 
-  const filteredSprints = sprints.filter(sprint => sprint.id == parameters.projectId);
+  const filteredSprints = sprints.filter(sprint => sprint.id === parameters.projectId);
 
-  const responseSprints = filteredSprints.map(sprint => {
+  return filteredSprints.map(sprint => {
```
Comment (Jdallos) [inline src/functions/list-project-sprints/handler.ts:35]: I meant implicit return with ()

File: src/libs/filter-utils.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
+export const filterByDate = (object: { start_date: string, end_date: string }, currentDate: Date, parameters: { startDate?: Date, endDate?: Date }): boolean => {
```
[the remark this answers is not in the document]
Comment (anttileppa) [reply src/libs/filter-utils.ts:1]: This should be wrapped inside a namespace

[the remark this answers is not in the document]
Comment (anttileppa) [reply src/libs/filter-utils.ts:1]: Add documentations

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: src/apis/forecast-api-service.ts [hunk 13]
```diff
@@ -2,47 +2,89 @@ import { Allocation } from "./schemas/allocation";
...
+}
+
 /**
- * Class for the Forecast API
+ * Creates ForecastApiService
  */
-export class ForecastApiService {
-    private apiKey: string;
+export function CreateForecastApiService(): ForecastApiService {
+    const apiKey: string = process.env.FORECAST_API_KEY;
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/apis/forecast-api-service.ts:18]: Task created:

File: src/apis/forecast-api-service.ts [hunk 14]
```diff
@@ -2,47 +2,89 @@ import { Allocation } from "./schemas/allocation";
 import { Project } from "./schemas/project";
 import { Task } from "./schemas/task";
 
+export interface ForecastApiService {
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/apis/forecast-api-service.ts:5]: Task created:
https://github.com/Metatavu/home-lambdas/issues/10

File: src/apis/schemas/sprint.ts [hunk 15]
```diff
@@ -0,0 +1,6 @@
+interface Sprint {
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/apis/schemas/sprint.ts:1]: Task created:
https://github.com/Metatavu/home-lambdas/issues/11

File: src/apis/schemas/task.ts [hunk 16]
```diff
@@ -7,5 +7,6 @@ export interface Task {
     estimate: number,
     start_date: string,
     end_date: string,
+    highPriority: boolean,
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/apis/schemas/task.ts:10]: Task created:
https://github.com/Metatavu/home-lambdas/issues/11

File: src/apis/schemas/time-entry.ts [hunk 17]
```diff
@@ -0,0 +1,7 @@
+interface TimeEntry {
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/apis/schemas/time-entry.ts:1]: Task created:
https://github.com/Metatavu/home-lambdas/issues/11

File: src/functions/list-allocations/handler.ts [hunk 18]
```diff
@@ -1,7 +1,73 @@
...
 // import { parseBearerAuth } from '@libs/auth-utils';
-import { middyfy } from '@libs/lambda';
-import { ForecastApiServiceFactory } from 'src/apis/forecast-api-service-factory';
+import { middyfy } from "@libs/lambda";
+import { CreateForecastApiService, ForecastApiService } from "src/apis/forecast-api-service";
+
+/**
+ * Parameters for lambda
+ */
+export interface ListAllocationsParameters {
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/functions/list-allocations/handler.ts:10]: Task created:
https://github.com/Metatavu/home-lambdas/issues/11

File: src/functions/list-allocations/handler.ts [hunk 19]
```diff
@@ -1,7 +1,73 @@
...
+  startDate?: Date,
+  endDate?: Date,
+  personId?: string,
+  projectId?: string,
+}
+
+/**
+ * Response schema for lambda
+ */
+export interface Response {
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/functions/list-allocations/handler.ts:20]: Task created:
https://github.com/Metatavu/home-lambdas/issues/11

File: src/functions/list-project-sprints/handler.ts [hunk 20]
```diff
@@ -0,0 +1,81 @@
+import { ValidatedEventAPIGatewayProxyEvent } from "@libs/api-gateway";
+// import { parseBearerAuth } from '@libs/auth-utils';
+import { middyfy } from "@libs/lambda";
+import { CreateForecastApiService, ForecastApiService } from "src/apis/forecast-api-service";
+
+/**
+ * Parameters for lambda
+ */
+export interface ListProjectSprintsParameters {
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/functions/list-project-sprints/handler.ts:9]: Task created:
https://github.com/Metatavu/home-lambdas/issues/11

File: src/functions/list-project-sprints/handler.ts [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+  projectId: number,
+}
+
+/**
+ * Response schema for lambda
+ */
+export interface Response {
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/functions/list-project-sprints/handler.ts:16]: Task created:
https://github.com/Metatavu/home-lambdas/issues/11

File: src/functions/list-projects/handler.ts [hunk 22]
```diff
@@ -1,7 +1,55 @@
...
 // import { parseBearerAuth } from '@libs/auth-utils';
-import { middyfy } from '@libs/lambda';
-import { ForecastApiServiceFactory } from 'src/apis/forecast-api-service-factory';
+import { middyfy } from "@libs/lambda";
+import { CreateForecastApiService, ForecastApiService } from "src/apis/forecast-api-service";
+
+/**
+ * Parameters for lambda
+ */
+export interface ListProjectsParameters {
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/functions/list-projects/handler.ts:10]: Task created:
https://github.com/Metatavu/home-lambdas/issues/11

File: src/functions/list-tasks/handler.ts [hunk 23]
```diff
@@ -1,40 +1,92 @@
...
+ * Parameters for lambda
+ */
+export interface ListTasksParameters {
+  projectId: number,
+}
+
+/**
+ * Response schema for lambda
+ */
+export interface Response {
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/functions/list-tasks/handler.ts:17]: Task created:
https://github.com/Metatavu/home-lambdas/issues/11

File: src/functions/list-tasks/handler.ts [hunk 24]
```diff
@@ -1,40 +1,92 @@
...
+import { ValidatedEventAPIGatewayProxyEvent } from "@libs/api-gateway";
+import { parseBearerAuth } from "@libs/auth-utils";
+import { middyfy } from "@libs/lambda";
+import { CreateForecastApiService, ForecastApiService } from "src/apis/forecast-api-service";
+import { Task } from "src/apis/schemas/task";
+
+/**
+ * Parameters for lambda
+ */
+export interface ListTasksParameters {
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/functions/list-tasks/handler.ts:10]: Task created:
https://github.com/Metatavu/home-lambdas/issues/11

File: src/functions/list-time-entries/handler.ts [hunk 25]
```diff
@@ -0,0 +1,81 @@
+import { ValidatedEventAPIGatewayProxyEvent } from "@libs/api-gateway";
+// import { parseBearerAuth } from '@libs/auth-utils';
+import { middyfy } from "@libs/lambda";
+import { CreateForecastApiService, ForecastApiService } from "src/apis/forecast-api-service";
+
+/**
+ * Parameters for lambda
+ */
+export interface ListTimeEntriesParameters {
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/functions/list-time-entries/handler.ts:9]: Task created:
https://github.com/Metatavu/home-lambdas/issues/11

File: src/functions/list-time-entries/handler.ts [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+  projectId: number
+}
+
+/**
+ * Response schema for lambda
+ */
+export interface Response {
```
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/functions/list-time-entries/handler.ts:16]: Task created:
https://github.com/Metatavu/home-lambdas/issues/11

File: src/libs/filter-utils.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (Jdallos) [reply src/libs/filter-utils.ts:1]: create an interface for the dates to describe their usage e.g. 'DateRange', rather than destructuring

**Pull request closed without merging**